### PR TITLE
[WIP] ADVENT conflict-solving branch

### PR DESCRIPTION
**Pull request by tianyu-z**

### CODE REVIEW

**Comment by tianyu-z** [comment]
I will request a review when I am ready (finishing the confliction and resolving the conversation)

**Comment by vict0rsch** [comment]
I know this is WIP but be careful: you have a weird `.bak` file which should not be committed and not pushed + you are overwriting the current `defaults.yaml` with an outdated one

**Comment by vict0rsch** [comment]
Also can you give this PR a more explicit name? I'm glad you're solving issues but that's not an ideal name haha

**Comment by tianyu-z** [comment]
> I know this is WIP but be careful: you have a weird `.bak` file which should not be committed and not pushed + you are overwriting the current `defaults.yaml` with an outdated one

Understood. That's why I decided to leave these 2 most important files until when I get up. (with a clear mind)

**Comment by tianyu-z** [comment]
> Also can you give this PR a more explicit name? I'm glad you're solving issues but that's not an ideal name haha

Got it!

**Comment by tianyu-z** [comment]
I think all the confliction has been solved. I also re-run an experiment [here](https://www.comet.ml/tianyu-z/omnigan/bc890405a0de400b8accc16eb75fed52).

I remembered @51N84D said he ran the omnigan-mask with the original size of the datasets. However, after setting these:
```
    - name: resize
      ignore: true
      new_size: 256
    - name: crop
      ignore: true
      height: 224
      width: 224
```
in the config file, I received an error message:
Traceback (most recent call last):
  File "train.py", line 135, in <module>
    trainer.setup()
  File "/network/home/tianyu.zhang/omnigan5/omnigan/trainer.py", line 172, in setup
    self.latent_shape = self.compute_latent_shape()
  File "/network/home/tianyu.zhang/omnigan5/omnigan/trainer.py", line 151, in compute_latent_shape
    b = Dict(next(iter(self.loaders[mode][domain])))
  File "/network/home/tianyu.zhang/.local/lib/python3.7/site-packages/torch/utils/data/dataloader.py", line 819, in __next__
    return self._process_data(data)
  File "/network/home/tianyu.zhang/.local/lib/python3.7/site-packages/torch/utils/data/dataloader.py", line 846, in _process_data
    data.reraise()
  File "/network/home/tianyu.zhang/.local/lib/python3.7/site-packages/torch/_utils.py", line 369, in reraise
    raise self.exc_type(msg)
RuntimeError: Caught RuntimeError in DataLoader worker process 0.
Original Traceback (most recent call last):
  File "/network/home/tianyu.zhang/.local/lib/python3.7/site-packages/torch/utils/data/_utils/worker.py", line 178, in _worker_loop
    data = fetcher.fetch(index)
  File "/network/home/tianyu.zhang/.local/lib/python3.7/site-packages/torch/utils/data/_utils/fetch.py", line 47, in fetch
    return self.collate_fn(data)
  File "/network/home/tianyu.zhang/.local/lib/python3.7/site-packages/torch/utils/data/_utils/collate.py", line 75, in default_collate
    return {key: default_collate([d[key] for d in batch]) for key in elem}
  File "/network/home/tianyu.zhang/.local/lib/python3.7/site-packages/torch/utils/data/_utils/collate.py", line 75, in <dictcomp>
    return {key: default_collate([d[key] for d in batch]) for key in elem}
  File "/network/home/tianyu.zhang/.local/lib/python3.7/site-packages/torch/utils/data/_utils/collate.py", line 75, in default_collate
    return {key: default_collate([d[key] for d in batch]) for key in elem}
  File "/network/home/tianyu.zhang/.local/lib/python3.7/site-packages/torch/utils/data/_utils/collate.py", line 75, in <dictcomp>
    return {key: default_collate([d[key] for d in batch]) for key in elem}
  File "/network/home/tianyu.zhang/.local/lib/python3.7/site-packages/torch/utils/data/_utils/collate.py", line 56, in default_collate
    return torch.stack(batch, 0, out=out)
RuntimeError: invalid argument 0: Sizes of tensors must match except in dimension 0. Got 512 and 3120 in dimension 2 at /pytorch/aten/src/TH/generic/THTensor.cpp:689

Do you have any suggestions? How did you train Omnigan on the original(un-resized) images? Thank you!

**Comment by tianyu-z** [comment]
I didn't integrate the ADVENTAdversarialLoss into the get_losses function because it's a little bit different. If you want me to put it into get_losses, just notify me.

**Comment by tianyu-z** [comment]
Advent dicription:
https://drive.google.com/file/d/13gbw5NNPANoWeFyIg7Pf8G4v_b8-53Az/view?usp=sharing

File: omnigan/losses.py [hunk 1]
```diff
@@ -284,4 +285,83 @@ def get_losses(opts, verbose):
...
+        
+        if self.opts.dis.ADVENT.multi_level == True:
+            d_out_aux = discriminator_aux(prob_2_entropy(F.softmax(prediction1, dim = 1)))
+            loss_adv_trg_aux = self.loss(d_out_aux, target)
+        else:
+            loss_adv_trg_aux = 0
+        d_out_main = discriminator_main(prob_2_entropy(F.softmax(prediction2, dim = 1)))
+        loss_adv_trg_main = self.loss(d_out_main, target)
+        
+        loss = (self.opts.train.lambdas.advent.adv_main * loss_adv_trg_main
```
Comment (vict0rsch) [inline omnigan/losses.py:364]: do the `lambdas` multiplications outside  the loss computation

File: omnigan/trainer.py [hunk 2]
```diff
@@ -573,6 +580,21 @@ def get_representation_loss(self, multi_domain_batch):
                         batch_domain
                     ] = update_loss.item()
 
+                    # Then ADVENT Fool loss
+                    if batch_domain == 'r':
+                        for param in self.D["m"]["maskAdvent"].parameters():
+                            param.requires_grad = False
+                        pred = prediction[:, 0, :, :]
+                        pred_prime = 1 - pred
+                        prob = torch.stack([pred, pred_prime]).transpose(0, 1)
```
Comment (vict0rsch) [inline omnigan/trainer.py:589]:
```
pred = prediction # not necessary, for the sake of the argument
pred_prime = 1 - pred
prob = torch.cat([pred, pred_prime], axis=1)
```

File: omnigan/trainer.py [hunk 3]
```diff
@@ -573,6 +580,21 @@ def get_representation_loss(self, multi_domain_batch):
...
+                        update_loss = self.losses["G"]["tasks"][update_task]["advent"](
+                            None, prob.to(self.device), self.source_label, 
+                            self.D["m"]["maskAdvent"]
+                        )
+                        self.logger.losses.task_loss[update_task]["advent"][
+                            batch_domain
+                        ] = update_loss.item()
```
Comment (vict0rsch) [inline omnigan/trainer.py:596]: add `step_loss`!

File: omnigan/trainer.py [hunk 4]
```diff
@@ -573,6 +580,21 @@ def get_representation_loss(self, multi_domain_batch):
...
+                        for param in self.D["m"]["maskAdvent"].parameters():
+                            param.requires_grad = False
+                        pred = prediction[:, 0, :, :]
+                        pred_prime = 1 - pred
+                        prob = torch.stack([pred, pred_prime]).transpose(0, 1)
+                        update_loss = self.losses["G"]["tasks"][update_task]["advent"](
+                            None, prob.to(self.device), self.source_label, 
+                            self.D["m"]["maskAdvent"]
+                        )
+                        self.logger.losses.task_loss[update_task]["advent"][
```
Comment (vict0rsch) [inline omnigan/trainer.py:594]: `losses.task_loss` => `losses.generator.task_loss`

Comment (vict0rsch) [reply omnigan/trainer.py:594]: Do that ☝️  in this whole section

File: omnigan/trainer.py [hunk 5]
```diff
@@ -796,38 +818,131 @@ def get_d_loss(self, multi_domain_batch, verbose=0):
...
+                    )
+                elif batch_domain == "s":
+                    disc_loss["m"]["maskAdvent"] += self.losses["D.m.advent"](
+                        None,
+                        prob.to(self.device),
+                        self.source_label,
+                        self.D["m"]["maskAdvent"]
+                    )
+                else:
+                    raise Exception("Wrong Domain Input!")
```
Comment (vict0rsch) [inline omnigan/trainer.py:872]: Remove that, not needed with `zerotensor`

File: omnigan/trainer.py [hunk 6]
```diff
@@ -796,38 +818,131 @@ def get_d_loss(self, multi_domain_batch, verbose=0):
...
+                        print(f"Batch {batch_domain} > {decoder}: {source_domain} to real ")
+                        print(f"Batch {batch_domain} > {decoder}: {target_domain} to fake ")
+            if "m" in self.opts.tasks:
                 if verbose > 0:
-                    print(f"Batch {batch_domain} > {decoder}: {source_domain} to real ")
-                    print(f"Batch {batch_domain} > {decoder}: {target_domain} to fake ")
+                    print("Now training the ADVENT discriminator!")
+
+                for param in self.D["m"]["maskAdvent"].parameters():
+                    param.requires_grad = True
```
Comment (vict0rsch) [inline omnigan/trainer.py:851]: move that outside the `update_d` call

File: omnigan/trainer.py (hunk 5 above)
Comment (tianyu-z) [reply omnigan/trainer.py]: Could you pls remind me what do I need to remove here?

File: omnigan/trainer.py [hunk 7]
```diff
@@ -767,9 +767,13 @@ def get_d_loss(self, multi_domain_batch, verbose=0):
         Returns:
             [type]: [description]
         """
+<<<<<<< Updated upstream
         zerotensor = torch.tensor(0).to(self.device)
+=======
+        zerotensor = torch.tensor(0.0).to(self.device)
+>>>>>>> Stashed changes
```
Comment (vict0rsch) [inline omnigan/trainer.py:774]: @tianyu-z you have this git marker trailing. Remember to run tests often! They wouldn't have passed

File: omnigan/trainer.py [hunk 8]
```diff
@@ -778,16 +782,21 @@ def get_d_loss(self, multi_domain_batch, verbose=0):
...
+                        torch.empty(
+                            batch_size,
+                            self.opts.gen.p.latent_dim,
+                            self.painter_z_h,
+                            self.painter_z_w,
+                        )
+                        .normal_(mean=0, std=1.0)
+                        .to(self.device)
                     )
+<<<<<<< Updated upstream
```
Comment (vict0rsch) [inline omnigan/trainer.py:799]: again

File: omnigan/trainer.py [hunk 9]
```diff
@@ -821,37 +845,44 @@ def get_d_loss(self, multi_domain_batch, verbose=0):
...
                         #     self.target_label,
-                        #     self.D["m"]["maskAdvent_aux"]
+                        #     self.D["m"]["Advent_aux"]
                         # )
 
-                        disc_loss["m"]["maskAdvent"] += (
+                        disc_loss["m"]["Advent"] += (
                             self.opts.train.lambdas.advent.adv_main * loss_main
                         )
+<<<<<<< Updated upstream
```
Comment (vict0rsch) [inline omnigan/trainer.py:860]: here too

File: omnigan/trainer.py [hunk 10]
```diff
@@ -821,37 +845,44 @@ def get_d_loss(self, multi_domain_batch, verbose=0):
...
-                        #     self.D["m"]["maskAdvent_aux"]
+                        #     self.D["m"]["Advent_aux"]
                         # )
 
-                        disc_loss["m"]["maskAdvent"] += (
+                        disc_loss["m"]["Advent"] += (
                             self.opts.train.lambdas.advent.adv_main * loss_main
                         )
                     else:
+<<<<<<< Updated upstream
```
Comment (vict0rsch) [inline omnigan/trainer.py:881]: and a last one :)

File: omnigan/trainer.py (hunk 7 above)
Comment (tianyu-z) [reply omnigan/trainer.py]: Thanks!

**Pull request merged**